### ISSUE TITLE
Unexpected behaviour for model.evaluate inside keras Callback

### ISSUE THREAD

**Issue by adriang133** [tensorflow/tensorflow#43262]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- TensorFlow installed from (source or binary): Colab
- TensorFlow version (use command below): 2.3.0
- Python version: 3.6.9

**Describe the current behavior**
Inside a `keras.callbacks.Callback`, `self.model.evaluate` returns results for the `validation_data` regardless of what is passed in. The issue seems to not be present if `validation_data` is None.

**Describe the expected behavior**
`self.model.evaluate` should evaluate what is passed in.

**Standalone code to reproduce the issue**
[Colab notebook
](https://colab.research.google.com/drive/1RHrGba3sQ3iw9Ju0_m3WRFC5dog4WG6k?usp=sharing) which illustrates the issue.

**Comment by bhack** [comment]
I think that problem is that with `def on_train_end` you are still in training mode.

**Comment by Saduf2019** [comment]
@adriang133 
Please update as per above comment

**Comment by adriang133** [comment]
I'm using `on_epoch_end` in my example. I'm not too familiar with the innards of Keras, but why would `on_epoch_end` have the model in training mode still ?

If this is intended, is there a way I can evaluate the model on a separate validation dataset at the end of each epoch ?

**Comment by bhack** [comment]
You can see what happens when you are calling `fit`: https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/engine/training.py#L1106

**Comment by adriang133** [comment]
I see. The model state during training changes if you specify `validation_data` vs if you don't which is unintuitive if you ask me, but if this is intended then feel free to close this issue.

TLDR for anyone else running into this, a solution is to move the evaluation of both validation datasets to the callback and call `fit(..., validation_data=None)`.

**Comment by bhack** [comment]
I don't know what you want to do but generally when you have something in `validation_data` you have also `on_train_begin` `on_train_end` callback available that you could customize.

**Comment by adriang133** [comment]
I want to evaluate the model on 2 (or more) validation sets at the end of each epoch. `on_train_end` is only called at the end of the training, not per epoch, so it wouldn't do. See #38803 for example. A google search yields a couple of similar solutions. They no longer work, maybe behaviour has changed ?  

As mentioned, this solution seems to work:

```python
class EvalCallback(keras.callbacks.Callback):
    def on_epoch_end(self, epoch, logs=None):
        for x, y in validation_datasets:
            print(self.model.evaluate(x, y, return_dict=True, verbose=0))

model.fit(..., validation_data=None, callbacks=[EvalCallback()])
```

Maybe Keras could provide a callback which evaluates one or more validation sets instead of the `validation_data` parameter?

**Comment by bhack** [comment]
If you want direct support for multiple validations we are going in the https://github.com/tensorflow/tensorflow/issues/38803 perimeter.
I think that you can close this and upvote that issue.
